tensorboard/uploader/uploader_test.py

```diff
@@ -50,65 +51,108 @@ class AbortUploadError(Exception):
-        stub = write_service_pb2_grpc.TensorBoardWriterServiceStub(test_channel)
-        mock_client = mock.create_autospec(stub)
-        fake_exp_response = write_service_pb2.CreateExperimentResponse(
-            experiment_id="123", url="should not be used!"
-        )
-        mock_client.CreateExperiment.return_value = fake_exp_response
-        return mock_client
+def _create_mock_client():
+    # Create a stub instance (using a test channel) in order to derive a mock
+    # from it with autospec enabled. Mocking TensorBoardWriterServiceStub itself
+    # doesn't work with autospec because grpc constructs stubs via metaclassing.
+    test_channel = grpc_testing.channel(
+        service_descriptors=[], time=grpc_testing.strict_real_time()
+    )
+    stub = write_service_pb2_grpc.TensorBoardWriterServiceStub(test_channel)
+    mock_client = mock.create_autospec(stub)
+    fake_exp_response = write_service_pb2.CreateExperimentResponse(
+        experiment_id="123", url="should not be used!"
+    )
+    mock_client.CreateExperiment.return_value = fake_exp_response
+    return mock_client
+
+
+_rpc_rate_limiter = util.RateLimiter(0)
 
+
+class TensorboardUploaderTest(tf.test.TestCase):
     def test_create_experiment(self):
         logdir = "/logs/foo"
-        mock_client = self._create_mock_client()
+        mock_client = _create_mock_client()
         uploader = uploader_lib.TensorBoardUploader(mock_client, logdir)
         eid = uploader.create_experiment()
         self.assertEqual(eid, "123")
 
     def test_start_uploading_without_create_experiment_fails(self):
-        mock_client = self._create_mock_client()
+        mock_client = _create_mock_client()
         uploader = uploader_lib.TensorBoardUploader(mock_client, "/logs/foo")
         with self.assertRaisesRegex(RuntimeError, "call create_experiment()"):
             uploader.start_uploading()
 
+    # Send each Event below in a separate WriteScalarRequest
+    @mock.patch.object(uploader_lib, "_MAX_REQUEST_LENGTH_BYTES", 100)
     def test_start_uploading(self):
-        mock_client = self._create_mock_client()
+        mock_client = _create_mock_client()
         mock_rate_limiter = mock.create_autospec(util.RateLimiter)
         uploader = uploader_lib.TensorBoardUploader(
             mock_client, "/logs/foo", mock_rate_limiter
         )
         uploader.create_experiment()
-        mock_builder = mock.create_autospec(uploader_lib._RequestBuilder)
-        request = write_service_pb2.WriteScalarRequest()
-        mock_builder.build_requests.side_effect = [
-            iter([request, request]),
-            iter([request, request, request, request, request]),
+
+        mock_logdir_loader = mock.create_autospec(logdir_loader.LogdirLoader)
+        mock_logdir_loader.get_run_events.side_effect = [
+            {
+                "run 1": [
+                    event_pb2.Event(
+                        summary=scalar_v2.scalar_pb("event 1.1", 5.0)
+                    ),
```

Comment:
Optional, but feel free to `def scalar_event(tag, value)` inside this
test function to save 20 lines of noise.